### DISCUSSION TITLE
SyntaxError: Unexpected end of JSON input at JSON.parse

### DISCUSSION THREAD

**Discussion by vladislav-gorovenko** [vercel/next.js#51011] [Help] [answered]
### Summary

Can't figure out why request.json() doesn't work in my code. Error is caused by` await request.json();` part. 

```
export async function DELETE(request: Request) {
  const { id }: Partial<Todo> = await request.json();

  if (!id) return NextResponse.json({ deleted: false });

  const res = await fetch(`${DATA_SOURCE_URL}/${id}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      "API-Key": API_KEY,
    },
  });

  const data = await res.json();

  return NextResponse.json({ ...data, deleted: true });
}
```

<img width="1094" alt="Screenshot 2023-06-09 at 11 14 23" src="https://github.com/vercel/next.js/assets/38355614/7ec127d7-2dc5-4734-8237-6c5d967b2933">

### Additional information

```js
- error SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6571:19)
    at successSteps (node:internal/deps/undici/undici:6545:27)
    at node:internal/deps/undici/undici:1211:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
- error SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6571:19)
    at successSteps (node:internal/deps/undici/undici:6545:27)
    at node:internal/deps/undici/undici:1211:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by icyJoseph** [comment]
Aha, you already did that part... Let's see... I think I've read about issues with the DELETE method before, I come back after some testing.

**Reply by vladislav-gorovenko** [reply]
Thank you! I saw this issue raised, but it was for previous versions of Next.js. I'm thinking about passing my id as a searchParam in a request url.

**Reply by icyJoseph** [reply]
Yeah, the thing I am actually testing is whether body in DELETE is also as "discouraged", as it is with GET, and HEAD.

**Reply by icyJoseph** [reply, marked as answer]
Alright, yeah it looks like it is murky waters, like with GET. It is not like explicitly allowed, but some implementations might not see it, or crash trying to parse it, while others just account for it.

**Reply by icyJoseph** [reply]
It is weird because DELETE is not explicitly mentioned here, https://developer.mozilla.org/en-US/docs/Web/API/Request/body:


> The read-only body property of the [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) interface contains a [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream) with the body contents that have been added to the request. Note that a request using the GET or HEAD method cannot have a body and null is returned in these cases.

**Reply by icyJoseph** [reply]
Sorry for spam, but at least it'd seem that the Node Request API is not the root issue here:

```console
 ~/dev-exp at  took 30s  ➜
node
Welcome to Node.js v18.14.1.
Type ".help" for more information.
> const request = new Request("https://example.com", { method: "DELETE", body: '{"foo": "bar"}', });
undefined
> await request.json();
{ foo: 'bar' }
```

🤔

**Reply by vladislav-gorovenko** [reply]
Hah, it's ok. I was thinking that I'm doing something wrong in my code, since I just started learning Next.js to build some pet projects for portfolio. I understand that it's an open question for now and will avoid using body for DELETE requests. Will pass all the parameters in url and will probable create a separate route. 

But yeah, it's confusing, since there is nothing mentioned on MDN whether I can use body or not and previous versions of Next.js worked perfectly fine with request.json() on DELETE.

**Reply by icyJoseph** [reply]
It's cool. Don't forget to mark as answered, if you think it is answered, at least partially for now...

**Reply by dwolrdcojp** [reply]
Hey just wanted to add to this as it helped resolve my issue. 

```bash
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6662:19)
    at successSteps (node:internal/deps/undici/undici:6636:27)
    at node:internal/deps/undici/undici:1236:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```
    
Was being caused by passing method = "DELETE" explicitly in my fetch wrapper
   
```javascript
     delete: async function (
    path: ZigEndpoints,
    options: ZigOptions = {},
    body?: any,
    token: string | null | undefined = typeof window !== "undefined"
      ? use(TokenContext)
      : null,
  ) {
    options.method = "DELETE";
    return await authFetch(path, options, token);
  },
```

Commenting out this line made the error go away
  
  ```javascript
      options.method = "DELETE";
```


Thanks.
